models/solution/route.rb

```diff
@@ -69,6 +97,16 @@ def shift_route_times(shift_amount, shift_start_index = 0)
             waiting_resorb = [stop.info.waiting_time - current_shift, 0].max
             current_shift = [current_shift - waiting_resorb, 0].max
             stop.info.waiting_time -= waiting_resorb
+          else
+            active_tw = stop.active_timewindow
+            if active_tw
+              current_shift =
+                if shift_amount >= 0
+                  [current_shift, active_tw.safe_end(stop.activity.lateness_allowed?) - stop.info.begin_time].min
+                else
+                  [current_shift, active_tw.start - stop.info.begin_time].max
```

Comment:
This review made me realize that with this "logic" we can lead to unnecessary idle time. Imagine a route without any idle time and we deleted a stop `s` -- so everything is shifted `s.duration` to the left. The first few services were able to shift to the left, okay but at some point, we hit a time window and shift amount became `0` -- i.e., we introduced intermediary idle times :clap: 

I don't have a solution.

As I said, when I first write this function, I only had the "no-TW" case in mind and it does the job for that case. However, when the time windows are involved we are not sure if it is better to shift everything following the deleted stop to the left or everything before it to the right or everything to the middle direction... I am afraid we are beating a dead horse. What do you think?

Reply:
I see two cases 

- Force start: We need to stick to the route begin_time
- Without force start, there is no need to shift the end of the route to the left, but to move the beginning of the route to the right

Perhaps shift route times and distance should take an interval of indices to update

Reply:
With the new case of removing the empty mission in first position, i can't be sure there is no timewindow :/

Reply:
I have no idea. This is getting more and more complicated. Maybe we should just delete the "empty" service and re-launch ortools? :D

Reply:
> Without force start, there is no need to shift the end of the route to the left, but to move the beginning of the route to the right

Even without force_start in the presence of tw.end, we might want to push left :/

Reply:
Do you have a concrete example/instance that we can launch to observe the use of unnecessary (and clearly sub-optimal) empty service as the first stop? Maybe we can fix that?